### ISSUE TITLE
Cannot add Haskell Stack SDK

### ISSUE THREAD

**Issue by c4dr01d** [rikvdkleij/intellij-haskell#450]
I installed the stack only at my windows pc, than I configure Haskell SDK, the idea says: is not invalid stack binary.
I don't know why.

**Comment by rikvdkleij** [comment]
Can you describe what you did?

**Comment by c4dr01d** [comment]
@rikvdkleij
1. Creating a new project
2. Configure Haskell SDK
3. Choose Stack binary executable.
After the idea report invalid stack binary error

**Comment by rikvdkleij** [comment]
Weird, the plugin checks trivial things like:
 - name contains `stack`
 - is not a directory
- stack it's version can be retrieved by option `--numberic-version`

**Comment by c4dr01d** [comment]
@rikvdkleij  the idea is report --numberic-version options error

**Comment by rikvdkleij** [comment]
Sorry, I mean, this command works:
`stack --numeric-version`

**Comment by c4dr01d** [comment]
@rikvdkleij  emm, ok. I will post the idea log at 7 O'clock, maybe can help to solve this problem.

**Comment by c4dr01d** [comment]
@rikvdkleij  This is error copied form idea log:
The selected file is not valid a stack binary.

**Comment by rikvdkleij** [comment]
I do not see a idea log.

**Comment by c4dr01d** [comment]
@rikvdkleij  The log just one line: The selected file is not valid a stack binary.
Others is running normally.

**Comment by rikvdkleij** [comment]
@Sir4ur0n Do you also have this on Windows?

**Comment by rikvdkleij** [comment]
In which folder did you install stack?

**Comment by c4dr01d** [comment]
@rikvdkleij  D:\Stack

**Comment by Sir4ur0n** [comment]
I'm in holidays and don't have access to my Windows machine, I can test in a few weeks if you're still interested

**Comment by c4dr01d** [comment]
@Sir4ur0n You stack.exe can identify by plugin?

**Comment by rikvdkleij** [comment]
@c4dr01d I have some idea which could fix this issue. I will let you know when a new version is available so you can test if it works for you.

**Comment by c4dr01d** [comment]
@rikvdkleij  Thanks

**Comment by Sir4ur0n** [comment]
Hi,

I just tested with Stack 2.1.3 and Intellij Haskell 1.0.0-beta51:
* Create new project from Intellij
* Select Stack binary
* Everything works fine

So I could not reproduce the issue.
Have you correctly installed Stack for Windows? FYI my Stack location is `C:\Users\sir4ur0n\AppData\Roaming\local\bin\stack.exe`

**Comment by rikvdkleij** [comment]
Should hopefully be solved in beta52.

**Comment by c4dr01d** [comment]
@rikvdkleij Hopefully this issue can be solved at new version, thanks.

**Comment by rikvdkleij** [comment]
@c4dr01d Can you check it works in beta52?

**Comment by rikvdkleij** [comment]
Please reopen if still an issue.

**Comment by rausm** [comment]
If you are on Windows, have valid stack binary (ie you can run stack new / init / built / <whatever> ), and stack --numeric-version returns sensible version, it might be Windows Defender - or other antivirus - influence.

Btw, when setting the SDK fails, and error message is written to Idea message/ error log (accessible in right bottom of startup dialog or editor window). In my case it contained "Access violation in generated code when reading 0000000094cdeebe" after the version number, which tipped me off. The weird thing is, in command line it worked properly / it only failed when launched from Idea.

**Comment by rikvdkleij** [comment]
@rausm Thanks for your comment! 

So it works for you or not? In the past I looked to this issue and I have no idea why setting the SDK does not always works on Windows.

**Comment by TranHHoang** [comment]
I solved this just by running IntelliJ as administration and everything works great!

**Comment by zored** [comment]
Another solution is to right-click on `stack.exe` in Explorer and grant it full access for users (I belive that there is sone command for that).

**Comment by rausm** [comment]
> So it works for you or not?
It started working (in IntelliJ) after I disabled Defender,
It worked Ok in plain command line.

**Comment by rubin55** [comment]
Just downloaded the current version of intellij-haskell (1.0.0-beta52) and got this issue also. The error message is:

```
Executing `C:\Users\rubin\.stack\bin\stack.exe --numeric-version` failed: C:\Users\rubin\.stack\bin\stack.exe --numeric-version: 2.1.3 Access violation in generated code when reading 0000000094cdeebe
```

**Comment by rubin55** [comment]
> 
> 
> Another solution is to right-click on `stack.exe` in Explorer and grant it full access for users (I belive that there is sone command for that).

I tested this, this does not work; Adding Users in the Security pane and granting full control does not change the behaviour - I still get the error.

**Comment by rubin55** [comment]
@rikvdkleij Can you re-open this issue? Will try and find out what's going on exactly. Running stack.exe with adminitrative privileges is not a valid workaround imho..

**Comment by rubin55** [comment]
I've noticed I can reproduce the issue from command-line, when setting compatibility mode to enabled and selecting any of ["Windows 8", "Windows 7", Windows Vista *"]. The Error message on the console is then exactly the same:

Windows 10 regular, no compatibility mode enabled:
```
C:\Users\rubin\.stack\bin>stack --numeric-version
2.1.3
```
Any compatibility mode enabled:
```
C:\Users\rubin\.stack\bin>stack --numeric-version
2.1.3
Access violation in generated code when reading 0000000094cdeebe

C:\Users\rubin\.stack\bin>echo %ERRORLEVEL%
1
```

**Comment by rubin55** [comment]
I've discovered this pull on vscode-hie-server: https://github.com/alanz/vscode-hie-server/pull/136, which explicitly unsets __COMPAT_LAYER before calling the stack.exe executable. I've verified that this indeed does work, but I can't figure out what that variable is set to when intellij-haskell calls stack.exe.

More details also on this thread, and comments:  https://github.com/commercialhaskell/stack/issues/3765#issuecomment-437823698

**Comment by rubin55** [comment]
So I redownloaded stack.exe from the gitub releases page of the stack project and wrote a small wrapper batch script that could either unset or echo __COMPAT_LAYER:

```
@echo off
set STACK_EXECUTABLE=stack-2.1.3-windows-x86_64.exe
echo __COMPAT_LAYER is set to: %__COMPAT_LAYER% > C:\Users\rubin\Desktop\out.txt
%STACK_EXECUTABLE% %*
```
I then added `stack.bat` as a stack executable. However, after doing this, the error did not occur anymore, which surprised me, to say the least. I then added the binary directly, where the only difference was now its name, and it added without failure 0_o?!

I then verified the hashes of the two stack executables:

```
C:\Users\rubin\.stack\bin>sha1sum stack-2.1.3-windows-x86_64.exe
97bbc809106fa330d5c2f9cd940a45023f7d65ff  stack-2.1.3-windows-x86_64.exe
C:\Users\rubin\.stack\bin>sha1sum stack.exe
97bbc809106fa330d5c2f9cd940a45023f7d65ff  stack.exe
```

As you can see, the previously used executable is identical. I then threw away the original `stack.exe`, copy'ed the just downloaded identical ` stack-2.1.3-windows-x86_64.exe`  to `stack.exe` and retried. Again, add without failure.

Long story short: I cannot reproduce after redownloading stack from the Stack projects' github releases page. I know __COMPAT_LAYER affects this behaviour and I strongly suspect Windows Defender is an actor in this somehow. I hope all the above helps other fellows who are using Stack and intellij-stack on Windows.

**Comment by iainnicol** [comment]
Inspired by the research of @rubin55 and others, I took a look at this.

Process Monitor from Sysinternals tells us stack is being spawned with environmental variable ``__COMPAT_LAYER=DetectorsAppHealth``. This is value beyond not documented: it returns one only useless search result.

I'm not certain it's Windows Defender setting this, but if not it definitely sounds it is caused by some similar Microsoft service for system health and telemetry. Indeed, using IntelliJ's ``Run anything`` command to run ``stack.exe --numeric-version`` I can reproduce the crash. But if I use the ``Run anything`` command to start a powershell session, then start stack from the powershell session, I cannot reproduce. This, to me, definitely sounds like an operating system level component is intercepting and modifying spawned processes.

Now, as to the nature of the crash when this environmental variable is set. Well, I'll leave that for commercialhaskell/stack#3765
